### PR TITLE
pluto: configure backups

### PR DESCRIPTION
**Pull request by mweinelt**
- sets the mountpoints for prometheus and victoriametrics dataset
- copies the backup module from non-critical-infra
- prepare the backup to the storagebox
- configure backups to include prometheus, grafana, victoriametrics

### CODE REVIEW

File: delft/pluto/default.nix [hunk 1]
```diff
@@ -22,6 +26,18 @@
     hostId = "e4c9bd10";
   };
 
+  age.secrets.pluto-ssh-key.file = ../secrets/pluto-ssh-key.age;
```
[the remark this answers is not in the document]
Comment (mweinelt) [reply delft/pluto/default.nix:29]: Sure, but how many ssh keypairs do we really need for a single host?

[the remark this answers is not in the document]
Comment (delroth) [reply delft/pluto/default.nix:29]: They don't really cost anything, do they? :p

I don't feel super strongly about this, it just feels weird to me to have multi purpose credentials if we don't *need* to.

File: delft/pluto/nixos-metrics.nix [hunk 2]
```diff
@@ -30,6 +30,21 @@
     timerConfig.OnCalendar = "12:00:00";
   };
 
+  services.backup = {
+    preHook = ''
+      ${config.boot.zfs.package}/bin/zfs snapshot -r ${config.fileSystems."/var/lib/victoriametrics".device}@backup
```
[the remark this answers is not in the document]
Comment (mweinelt) [reply delft/pluto/nixos-metrics.nix]: Pushed a draft of the module changes.

File: modules/backup.nix [hunk 3]
```diff
@@ -1,10 +1,24 @@
...
+, pkgs
 , ...
 }:
 
 let
   cfg = config.services.backup;
+
+  mkZfsPreHook = mountpoint: ''
+    DATASET="$(findmnt -nr -o -target -T "${mountpoint}")"
+    zfs snapshot -r "$DATASET@borg"
```
Comment (delroth) [inline modules/backup.nix:12]: Can we make this loudly fail if the snapshot already exists to avoid any potential race conditions?

Comment (delroth) [reply modules/backup.nix:12]: (It might already, but I'm not sure.)

File: modules/backup.nix [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+        Paths to include in the backup.
+      '';
+    };
+    includesZfsDatasets = mkOption {
+      type = listOf str;
+      default = [];
+      description = ''
+        ZFS datasets referenced by mountpoint to snapshot and include
+      '';
+    };
```
Comment (mweinelt) [inline modules/backup.nix:94]: Not too happy about the name. Taking the mountpoint is necessary to declaratively include it in the backup paths. I'm resolving the device name using `findmnt`.

**Review by delroth — APPROVED** [review]
Everything I've got remaining on this PR is basically small nits, feel free to address or dismiss at your convenience :-)

File: modules/backup.nix (hunk 3 above)
Comment (mweinelt) [reply modules/backup.nix:12]: It already will.

```
 borgbackup-job-state-start[456920]: cannot create snapshot 'zroot/root/prometheus@backup': dataset already exists
 borgbackup-job-state-start[456920]: no snapshots were created
 borgbackup-job-state-start[456926]: could not find any snapshots to destroy; check snapshot names.
 systemd[1]: borgbackup-job-state.service: Main process exited, code=exited, status=1/FAILURE
 systemd[1]: borgbackup-job-state.service: Failed with result 'exit-code'.

File: delft/pluto/disko.nix [hunk 5]
```diff
@@ -83,6 +83,14 @@
             type = "zfs_fs";
             mountpoint = "/";
           };
+          "root/prometheus" = {
```
[the remark this answers is not in the document]
Comment (mweinelt) [reply delft/pluto/disko.nix:86]: With the backup module handling the abstraction, how do you feel about keeping it this way?

**Comment by mweinelt** [comment]
Untested.

**Comment by mweinelt** [comment]
Deployed and tested. I think the zfs integration in the backup module provides sufficient decoupling for now.

**Pull request merged**